crates/worker/src/partition_processor_manager.rs

```diff
@@ -581,97 +634,142 @@ impl<T: TransportConnect> PartitionProcessorManager<T> {
             CreateSnapshot(partition_id, sender) => {
                 self.running_partition_processors
                     .get(&partition_id)
-                    .map(|store| store.handle.create_snapshot(Some(sender)));
+                    .map(|store| {
+                        let ProcessorStatus::Started(state) = store else {
+                            return None;
+                        };
+
+                        Some(state.handle.create_snapshot(Some(sender)))
+                    });
             }
             GetState(sender) => {
                 let _ = sender.send(self.get_state());
             }
         }
     }
 
-    async fn on_control_processors(
+    fn on_control_processors(
         &mut self,
         control_processor: Incoming<ControlProcessors>,
-    ) -> Result<(), Error> {
+        effects: mpsc::UnboundedSender<ManagerEffect>,
+    ) {
         let control_processors = control_processor.into_body();
 
-        self.metadata
-            .wait_for_version(
-                MetadataKind::Logs,
-                control_processors.min_logs_table_version,
-            )
-            .await?;
-        let partition_table = self
-            .metadata
-            .wait_for_partition_table(control_processors.min_partition_table_version)
-            .await?
-            .into_arc();
-
-        for control_processor in control_processors.commands {
-            self.on_control_processor(control_processor, &partition_table)
+        let metadata = self.metadata.clone();
+        let handle_control_processors = async move {
+            metadata
+                .wait_for_version(
+                    MetadataKind::Logs,
+                    control_processors.min_logs_table_version,
+                )
                 .await?;
-        }
 
-        Ok(())
+            let partition_table = metadata
+                .wait_for_partition_table(control_processors.min_partition_table_version)
+                .await
+                .map(|pt| pt.into_arc())?;
+
+            for control_processor in control_processors.commands {
+                effects.send(ManagerEffect {
+                    parent: Span::current(),
```

Comment:
I was thinking that we should include the versions here to avoid race between multiple control processors messages, by filtering on the version before applying the effect

Reply:
I think this can happen and we need to handle it.

Probably it is enough to only apply the latest of the `ControlProcessors` message because by the time we have received the updated metadata, the cluster controller could have made a different decision.

So what we could do is to remember the latest `ControlProcessors` message in an `Option<ControlProcessors>` field. Then we could have something like:

```
let mut partition_table_version_watcher = metadata.watch(MetadataKind::PartitionTable);
let mut logs_version_watcher = metadata.watch(MetadataKind::Logs);

tokio::select! {
    ...
    _ = partition_table_version_watcher.changed(), if self.control_processors.is_some() => {
        // check whether we can apply the control_processors message
    }
    _ = logs_version_watcher.changed(), if self.control_processors.is_some() => {
        // check whether we can apply the control_processors message
    }
    ...
}
```

in the main event loop to await the version update. At the same tiem we can update `self.control_processors` field with newer control instructions.

Reply:
That's a very good approach and is definitely better than tracking versions. Thank you!